compiler/rustc_codegen_llvm/src/consts.rs

```diff
@@ -282,6 +283,12 @@ impl CodegenCx<'ll, 'tcx> {
             }
         }
 
+        if self.tcx.sess.relocation_model() == RelocModel::Static {
```

Comment:
Can this (re-)use `should_assume_dso_local`?

Reply:
Tbh initially I've just copied the check from [here](https://github.com/rust-lang/rust/blob/0c7d4effd71f826abb9651419c316b9b57b8ac97/compiler/rustc_codegen_llvm/src/callee.rs#L184-L186) since we don't directly have the visibility and linkage available. But I'm able query those using `LLVMRustGetLinkage`/`LLVMRustGetVisibility`.

The problem is if I use `should_assume_dso_local` [this test](https://github.com/rust-lang/rust/blob/75da570d784a798a34ff1e5048cd9a6a2fb23170/src/test/ui/threads-sendsync/thread-local-extern-static.rs) (and apparently only this single one) fails with
```
/usr/bin/ld: /home/bobo1239/Development/Rust/rust/build/x86_64-unknown-linux-gnu/test/ui/threads-sendsync/thread-local-extern-static/a.thread_local_extern_static.7rcbfp3g-cgu.0.rcgu.o(.text._ZN26thread_local_extern_static4main17h668b1926d155f877E+0x8): unresolvable R_X86_64_TPOFF32 relocation against symbol `FOO'
```
`FOO` is marked as `dso_local` in the LLVM IR since we run into [this branch](https://github.com/rust-lang/rust/blob/0c7d4effd71f826abb9651419c316b9b57b8ac97/compiler/rustc_codegen_llvm/src/mono_item.rs#L118-L121) but I don't understand relocations well enough to tell what is wrong here exactly. (whether setting `dso_local` is wrong here or LLVM does strange things) Without `dso_local` the compiler currently generates a `R_X86_64_GOTTPOFF` (vs. `R_X86_64_TPOFF32` with `dso_local`) relocation for `FOO`. The auxiliary library (which gets linked in) stays the same so the only difference should be the relocation type.

Reply:
I wonder if this is occurring because of multiple codegen units or something. The relocation kind doesn't look wrong in this instance, but I also just started figuring out them relocations myself, so :shrug:.

Reply:
Interestingly the final linking only fails when the [auxiliary crate](https://github.com/rust-lang/rust/blob/75da570d784a798a34ff1e5048cd9a6a2fb23170/src/test/ui/threads-sendsync/auxiliary/thread-local-extern-static.rs) is built with `--crate-type dylib` (keeping the rest of the flags the same as during test). Not sure what that means though...

Reply:
Hm, maybe it does make sense that it fails since with `--crate-type dylib` we're producing a `.so` for the auxiliary which is (I believe) a dso on it's own. So I guess we shouldn't be setting `dso_local` for [`FOO`](https://github.com/rust-lang/rust/blob/75da570d784a798a34ff1e5048cd9a6a2fb23170/src/test/ui/threads-sendsync/thread-local-extern-static.rs#L16) in the dependant crate. At least when we're linking with another dso.
Does this make sense? And is there a way to determine what we're linking with?

Reply:
Ah, I see where the problem is. `should_assume_dso_local` is only valid for definitions, not declarations (because those can be resolved by external DSOs). This particular piece of code is dealing with declarations (is inside a `!def_id.is_local()` conditional) so "just" a check with the relocation model is fine here (though it would be great to abstract this away somehow, still).  The other piece of code that copies over the `dso_local` flag, on the other hand *is* defining a static so I _think_ adding a `dso_local` flag there based on check from `should_assume_dso_local` should be valid.

Reply:
That makes sense to me except for the "dealing with declarations (is inside a `!def_id.is_local()` conditional)" part. The if block already ends some lines above unless I'm misunderstanding something.

Reply:
Ah, well, my bad. Reading code on Github sucks :( Still, the code deals with declarations _too_.